py: Fix unwinding of try-finally with break in finally clause

dpgeorge opened the pull request:
This set of patches fixes a bug in the VM when breaking within a try-finally, and also follows up with optimisations that open up with the bug fixed.

The bug has to do with executing a break within the finally block of a try-finally statement.  For example:
```python
def f():
    for x in (1,):
        print('a', x)
        try:
            raise Exception
        finally:
            print(1)
            break
        print('b', x)
f()
```
Currently in uPy the above code will print:
```
a 1
1
1
segmentation fault (core dumped)  micropython
```
Not only is there a seg fault, but the "1" in the finally block is printed twice.  This is because when the VM executes a finally block it doesn't really know if that block was executed due to a fall-through of the try (no exception raised), or because an exception is active.  In particular, for nested finallys the VM has no idea which of the nested ones have active exceptions and which are just fall-throughs.  So when a break (or continue) is executed it tries to unwind all of the finallys, when in fact only some may be active.

It's questionable whether break (or return or continue) should be allowed within a finally block, because they implicitly swallow any active exception, but nevertheless it's allowed by CPython (although almost never used in the standard library).  And uPy should at least not crash in such a case.

It took me a while to find an acceptable solution to this bug.  Initially I thought to just ban such a construct in the compiler, but eventually I came across the solution presented here.  It relies on the fact that exception and finally handlers always appear in the bytecode after the try body.

Note: there was a similar bug with a return in a finally block, but that was previously fixed in b735208403a54774f9fd3d966f7c1a194c41870f

----

To understand the fix it's best to look at each commit here independently.  In the end the following statements can be made:

1. The bug with break/continue in (nested) finallys is fixed.

2. The opcodes POP_BLOCK and POP_EXCEPT are replaced by POP_EXCEPT_JUMP.

3. Generated bytecode is reduced by 2 bytes for each use of a try-except (due to point 2 above).

4. Code size is reduced:
```
   bare-arm:   -88 -0.133% 
minimal x86:  -200 -0.129% 
   unix x64:  -320 -0.065%
unix nanbox:  -412 -0.093%
      stm32:  -192 -0.052% 
     cc3200:  -128 -0.069% 
    esp8266:  -676 -0.104% 
      esp32:  -444 -0.040%
```

dpgeorge committed:
py: Fix bug with unwinding of try-finally.

dpgeorge committed:
py/vm: Remove currently_in_except_block variable.

No longer needed.

dpgeorge committed:
py: Replace POP_BLOCK and POP_EXCEPT opcodes with POP_EXCEPT_JUMP.

POP_BLOCK and POP_EXCEPT are now the same, and are always followed by a
JUMP.  So this optimisation reduces code size, and RAM usage of bytecode by
two bytes for each try-except handler.

pfalcon commented on tests/basics/try_finally_break.py:
```diff
@@ -0,0 +1,76 @@
+                try:
+                    raise Exception
+                finally:
+                    break
```
Wouldn't it be interesting to add here more "finally" configuration cases, e.g. mix of where sub-finally is nested within parent's "finally" vs parent's body?

pfalcon replied:
We also don't have a single test for try/except/else in the testsuite.

dpgeorge replied:
Yes it would be good to add both these kinds of tests.  The former can go in this set of patches here.  The latter (else) would be a separate thing.

pfalcon replied:
Bottom line, is that merging this without testing the `else:` clauses shouldn't be done to avoid surprises.

pfalcon commented on py/showbc.c:
```diff
@@ -401,14 +401,9 @@ const byte *mp_bytecode_print_str(const byte *ip) {
-            printf("POP_EXCEPT");
+        case MP_BC_POP_EXCEPT_JUMP:
+            DECODE_ULABEL; // these labels are always forward
+            printf("POP_EXCEPT_JUMP " UINT_FMT, (mp_uint_t)(ip + unum - mp_showbc_code_start));
```
printf() of course not supposed to go in master. And likely the reason why minimal port grew whole 80 bytes.

dpgeorge replied:
It's just replacing one printf with another; this is not compiled into minimal.

pfalcon replied:
Oops sorry, missed this was showbc.c, thought it was vm.c.

dpgeorge commented:
On my local machine minimal x86 (`make`) decreased by 200 bytes, and minimal thumb (`make CROSS=1`) decreased by 104 bytes.  So there's something wrong with the Travis size check.

dpgeorge commented:
> Bottom line, is that merging this without testing the else: clauses shouldn't be done to avoid surprises.

Agreed.  I've now push some try-except-else(-finally) tests to master, rebased this on top of that, and added a test here for try-except-else-finally-break.

pfalcon commented:
Thanks.

So, is it the case that MicroPython never used POP_BLOCK to implement for/while loops, and resolved break/continue to direct jumps? I missed that. Something they go to in CPython3.8 only ;-).

With that in mind intuitive concerns of "why it's different in CPython" should be silenced.

dpgeorge commented:
> So, is it the case that MicroPython never used POP_BLOCK to implement for/while loops, and resolved break/continue to direct jumps?

Correct.  The direct jumps are unwinding jumps that call any outstanding finally blocks.  So the POP_BLOCK is implicit in the unwind jump, I guess you could say.

> With that in mind intuitive concerns of "why it's different in CPython" should be silenced.

At the very start of this project I tried to match CPython opcode-for-opcode in the generated bytecode, to make sure I didn't miss anything subtle (hence the opcode names sometimes don't make sense because they corresponded to CPython's names).  But since then uPy has moved on, so has CPython, and they now diverge a fair bit.

pfalcon reviewed (approved):
Please merge.

dpgeorge commented:
Merged in e1fb03f3e2d1e99d0f745e9e17a963a036d34476 through 5a2599d96299ad37cda954f1de345159f9acf11c
